Feature/inventory complete

Tides opened the pull request:
Opening up this PR early for you guys to review if y'all want to I know there's alot of to see here 😅
I'm also gonna be listing features that are done/being worked on still

- [x] Persistent inventory (in-mem for now)
- [x] Declaring/initializing recipes in the recipe book
- [x] Recipe builders for all the recipe types
- [ ] All inventory actions (drag, double, shift clicking etc..)
- [ ] Working crafting... (includes smelting, stonecutting and smithing)

Here's also a list of ones that I think would be good to look over:

1. 6be1502
2. 7187766
3. 5dfd8cf

Tides committed:
New inventory tingz

Tides committed:
Update more inventory stuff

Tides committed:
Made ItemMeta a struct again and removed count from it

Decided its probably better to just keep an array of ItemStack instead of micro optimizing

Tides committed:
Some updates

Renamed ItemMeta.Description to ItemMeta.Lore and made it a list and updated WriteSlot to properly write in lore/names.

Tides committed:
Updated nbt writer to include extra trags

Tides committed:
Created block meta related stuff and updated ItemMeta related stuff

Tides committed:
Chunks now store block metadata and made it possible to pull/set them from World class

Tides committed:
Merge branch 'master' into feature/inventory-complete

Tides committed:
Sound and position changes

Tides committed:
More inventory related stuff

Tides committed:
Merge branch 'master' into feature/inventory-complete

Tides committed:
Added all the item/block types

Tides committed:
Fixed up slot writing/reading

Tides committed:
Added some checks and store Inventory location

Tides committed:
Close window and make inventory semi-persistent

Tides committed:
Implement Recipe classes

Tides committed:
Added declare recipe packet and made a recipe registry

Tides committed:
Merge branch 'master' into feature/inventory-complete

Tides committed:
Updated recipes json with valid data

Tides committed:
renamed RecipeItem to ingredient and removed ID from interface

Tides committed:
Fixed up declare recipe serialization

Tides committed:
Unlock recipe and some more inventory stuff

Tides committed:
Space

Tides committed:
Added support for all interactable blocks and cleaned up inventory related packets

Jonpro03 committed:
Two steps forward, 10 steps back
impl craft recipe request/response

Jonpro03 committed:
Fix bug where recipies weren't sent correctly

Tides committed:
Ingredient refactor

Tides committed:
Update Client.cs

Tides committed:
Fix tags not writing out properly

Tides committed:
Final fix for ingredient converter.

Added 2 new methods to registry for an easier time getting a itemstack with a count of 1

Tides committed:
Recipe builders and a bit of refactors

Tides committed:
Showcasing the shaped builder :v

Tides committed:
Refactor namespace

Tides committed:
Updated recipe classes

Craftplacer reviewed (changes requested):
Would've been easier to review if you made large scale refactoring another PR, there are multiple things that are odd in this PR.

Tides committed:
Refactors

Moved CraftingType from api back to main as it won't be used by users.
Removed setting of empty strings when creating new inventories and

Tides committed:
Removed useless logging and gave an inventory method some description

Removed duplicate crafting type

Tides committed:
Update MinecraftStream.cs

Tides committed:
Update base recipe classes and builders

Tides committed:
Some refactors

Tides committed:
Final refactor :v
